### PR TITLE
ddl: set jobs dependency by schema and table name

### LINKED ISSUE #49498: TiCDC receive DDLs out of order by the DDL Job StartTs
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```
create table test.t;
drop table drop.t;
create table test.t;
```

There are 3 DDL jobs generated and synced to the TiCDC changefeed.

### 2. What did you expect to see? (Required)

The TiCDC changefeed should receive the DDL with small StartTs first, then the DDL with larger StartTs.  

### 3. What did you see instead (Required)

The TiCDC changefeed receives DDL with larger StartTs first, then the DDL with smaller StartTs.

### 4. What is your TiDB version? (Required)

`v6.5.3`

### PR DESCRIPTION
**Pull request by tangenta**
### What problem does this PR solve?

Issue Number: close #49498

Problem Summary:

We should also consider schema & table name before picking a DDL job to execute.

### What changed and how does it work?

- Extract `ddl.runningJobs` to a separate file.
- Add `AffectedSchemaNames` and `AffectedTableNames` to `model.Job`.
- Refine job depenency check.

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/ddl/ddl_api.go [hunk 1]
```diff
@@ -6898,6 +6924,9 @@ func (d *ddl) renameTables(ctx sessionctx.Context, oldIdents, newIdents []ast.Id
 		BinlogInfo: &model.HistoryInfo{},
 		Args:       []interface{}{oldSchemaIDs, newSchemaIDs, tableNames, tableIDs, oldSchemaNames, oldTableNames},
 		CtxVars:    []interface{}{append(oldSchemaIDs, newSchemaIDs...), tableIDs},
+
+		AffectedSchemaNames: affectedSchemaNames,
+		AffectedTableNames:  affectedTableNames,
```
Comment (zimulala) [inline pkg/ddl/ddl_api.go:6929]: Could we put these fields to `MultiSchemaInfo`?
There seems to be some overlap between `TableName` and `AffectedTableNames`.
But in that case, the `rename table` operation is a little tricky to explain. Or we can specially handle this operation by getting the `RawArgs` information.

Comment (tangenta) [reply pkg/ddl/ddl_api.go:6929]: No, 'MultiSchemaInfo' is only used for multi-schema changes, and this PR is not related to multi-schema changes.

`rename table` may change `RawArgs` during execution, with increasing complexity.

Comment (zimulala) [reply pkg/ddl/ddl_api.go:6929]: Or consider putting some related fields into a field of this Job.

**Review by zimulala — COMMENTED** [review]
Could we use `AffectedSchemaNames` to update `job2SchemaNames`？
https://github.com/pingcap/tidb/issues/43818

File: pkg/ddl/job_table.go [hunk 2]
```diff
@@ -138,10 +138,8 @@ func (d *ddl) getJob(se *sess.Session, tp jobType, filter func(*model.Job) (bool
 }
 
 func hasSysDB(job *model.Job) bool {
-	sNames := job2SchemaNames(job)
```
Comment (zimulala) [inline pkg/ddl/job_table.go:163]: I think we can remove `job2SchemaNames` after this update.

Comment (tangenta) [reply pkg/ddl/job_table.go:163]: Updated.

File: pkg/ddl/ddl_api_test.go [hunk 3]
```diff
@@ -145,3 +150,66 @@ func enQueueDDLJobs(t *testing.T, sess sessiontypes.Session, txn kv.Transaction,
...
+	for row := iter.Begin(); row != iter.End(); row = iter.Next() {
+		jobMeta := row.GetBytes(0)
+		job := model.Job{}
+		err = job.Decode(jobMeta)
+		require.NoError(t, err)
+		finishTSs = append(finishTSs, job.BinlogInfo.FinishedTS)
+	}
+	create1TS, dropTS, create0TS := finishTSs[0], finishTSs[1], finishTSs[2]
+	require.Less(t, create0TS, dropTS, "first create should finish before drop")
+	require.Less(t, dropTS, create1TS, "second create should finish after drop")
```
Comment (zimulala) [inline pkg/ddl/ddl_api_test.go:214]: Test failed：
```
 ddl_api_test.go:214: 
        	Error Trace:	pkg/ddl/ddl_api_test.go:214
        	Error:      	"446560642158297088" is not less than "446560641896677376"
        	Test:       	TestCreateDropCreateTable
```

File: pkg/ddl/ddl_running_jobs.go [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+	"strconv"
+	"strings"
+	"sync"
+
+	"github.com/pingcap/tidb/pkg/parser/model"
+)
+
+type runningJobs struct {
+	sync.RWMutex
+	ids           map[int64][]model.InvolvingSchemaInfo
```
Comment (Benjamin2037) [inline pkg/ddl/ddl_running_jobs.go:31]: Could we add another map(as revert index) using schema+tablen as the key and the "jobid1, jobid2" string as a value to make check conflict more efficient?

Comment (tangenta) [reply pkg/ddl/ddl_running_jobs.go:31]: OK.

File: pkg/ddl/ddl_api_test.go (hunk 3 above)
Comment (tangenta) [reply pkg/ddl/ddl_api_test.go:214]: The failpoint `mockOwnerCheckAllVersionSlow` does not work in time. I have changed the sleep time from 1 second to 2 seconds.

**Review by zimulala — APPROVED** [review]

File: pkg/ddl/ddl_api.go [hunk 5]
```diff
@@ -6849,6 +6875,10 @@ func (d *ddl) renameTable(ctx sessionctx.Context, oldIdent, newIdent ast.Ident,
 		BinlogInfo: &model.HistoryInfo{},
 		Args:       []interface{}{schemas[0].ID, newIdent.Name, schemas[0].Name},
 		CtxVars:    []interface{}{[]int64{schemas[0].ID, schemas[1].ID}, []int64{tableID}},
+		InvolvingSchemaInfo: []model.InvolvingSchemaInfo{
```
Comment (Benjamin2037) [inline pkg/ddl/ddl_api.go:6878]: whatif the rename statement operators multi-table？

File: pkg/ddl/ddl_api.go [hunk 6]
```diff
@@ -6898,6 +6934,8 @@ func (d *ddl) renameTables(ctx sessionctx.Context, oldIdents, newIdents []ast.Id
 		BinlogInfo: &model.HistoryInfo{},
 		Args:       []interface{}{oldSchemaIDs, newSchemaIDs, tableNames, tableIDs, oldSchemaNames, oldTableNames},
 		CtxVars:    []interface{}{append(oldSchemaIDs, newSchemaIDs...), tableIDs},
+
```
Comment (Benjamin2037) [inline pkg/ddl/ddl_api.go:6937]:
```suggestion
```

File: pkg/ddl/ddl_api.go [hunk 7]
```diff
@@ -5044,6 +5066,10 @@ func (d *ddl) ExchangeTablePartition(ctx sessionctx.Context, ident ast.Ident, sp
 		BinlogInfo: &model.HistoryInfo{},
 		Args:       []interface{}{defID, ptSchema.ID, ptMeta.ID, partName, spec.WithValidation},
 		CtxVars:    []interface{}{[]int64{ntSchema.ID, ptSchema.ID}, []int64{ntMeta.ID, ptMeta.ID}},
+		InvolvingSchemaInfo: []model.InvolvingSchemaInfo{
```
Comment (Benjamin2037) [inline pkg/ddl/ddl_api.go:5069]: I saw you have added one blank line since 6937,  please make it all consistent.

**Review by Benjamin2037 — APPROVED** [review]
LGTM

**Review by D3Hunter — APPROVED** [review]
parser part lgtm

**Pull request merged**